### ISSUE TITLE
Allow to configure if ArrayTrailingComma should ignore single-line arrays

### ISSUE THREAD

**Issue by jsotuyod** [checkstyle/checkstyle#3201]
Currently the [ArrayTrailingComma](http://checkstyle.sourceforge.net/config_coding.html#ArrayTrailingComma) automatically ignores inline declarations. So using a configuration such as:

``` xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ArrayTrailingComma"/>
    </module>
</module>
```

will never report on constructs such as:

``` java
public class Foo {
    public int[] getBar() {
        return new int[] { 0 };
        //                  ^ doesn't report missing trailing comma
    }
}
```

``` bash
$ java -jar ~/Downloads/checkstyle-6.18/checkstyle-6.18-all.jar -c config.xml Foo.java
Starting audit...
Audit done.
```

This behaviour is by design and is documented. However, it would be nice if such behaviour could be configured through a property (keeping the current behaviour as default for backwards compatibility), so that we may write a configuration such as:

``` xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ArrayTrailingComma">
            <property name="ignoreInlineArrays" value="false"/>
        </module>
    </module>
</module>
```

And have it report a missing trailing comma on the given code, forcing developers to write:

``` java
public class Foo {
    public int[] getBar() {
        return new int[] { 0, };
    }
}
```

Making the new `ignoreInlineArrays` property default to `true` would guarantee backwards compatibility, allowing users to opt-in to this behaviour.
##

**Comment by romani** [comment]
please report issue as described at http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug

**Comment by jsotuyod** [comment]
@romani updated, let me know if I'm missing anything. Also, did you know [you can provide templates for issues and PRs](https://github.com/blog/2111-issue-and-pull-request-templates) by adding some files under a `.github` directory?

**Comment by rupak0577** [comment]
I am on it.

**Comment by unrealwork** [comment]
I am on it.

**Comment by romani** [comment]
connection with @unrealwork  is lost, PR is closed as incomplete.
If smb want to continue from scratch or from where @unrealwork is finished, please be welcome .

**Comment by romani** [comment]
@jsotuyod ,

we got PullRequest with fix for your issue, so it bring my attention back to this issue.

can you provide more details on why it is reasonable to ignore single line array declarations ? We need to put reason in documentation.

we need to find better name for property as `ignore.....` usually mean completely skip from consideration but you do expect violation.

@kukreja-vikramaditya , can you suggest good name ? 

from documentation:
> The check demands a comma at the end if neither left nor right curly braces are on the same line as the last element of the array

so there should be violation on a case that @jsotuyod  is provided , it is even show in documentation .... strange ... looks like we have a bug.

**Comment by jsotuyod** [comment]
The whole point of having a trailing comma is to simplify reordering the array.

In general, and in particular under the current implementation, this is restricted to swapping lines in the IDE / cutting and pasting whole lines in a different order.

However, the cut and paste approach should still work for single line arrays. For instance, given the following array:

```java
int color[] = { BLACK, RED, BLUE, YELLOW, }
```

I can stand *at any one point*, and performing a 3 word cut, go paste it at a different position and the result will be consistent (for all VIM fans out there `3ciw`!). This wouldn't work without the trailing comma after `YELLOW`.

As for the name, I suggested `ignoreInlineArrays` default to true so as to completely ignore them by default (backwards-compatible), and I would only expect a violation when set to `false`. I believe this is right according to your own reasoning.

**Comment by romani** [comment]
@jsotuyod , thanks for explanation.

> I believe this is right according to your own reasoning.

yes, thanks for correcting me.

I have some concerns about name of property - `...Inline...`.
To my mind it is better to use smth like `ignoreSameLineDeclaration` ..... but in result and consider your reason ... Why we so focused on single line arrays ?
If follow you reasoning you always want to have comma in array, examples:
```
$ cat MyClass.java 
public class MyClass {
    public int[] getBar() {
        return new int[] { 0 };
        return new int[] { 0, 1 };
        return new int[]
                         { 0 };
        return new int[]
                         { 0, 1 };
        return new int[]
                         { 0, 1,
                           2, 3 // now violation is only there
                         };
    }
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ArrayTrailingComma">
        </module>
    </module>
</module>

$ java -jar checkstyle-8.1-all.jar -c config.xml MyClass.java 
Starting audit...
[ERROR] /var/tmp/MyClass.java:12: Array should contain trailing comma. [ArrayTrailingComma]
Audit done.
Checkstyle ends with 1 errors.
```

and by your reasoning/logic you will expect violation in all that cases if some option is use, am I right ?
if yes, so we can have property `alwaysDemandTrailingComma`

**Comment by jsotuyod** [comment]
@romani yes, definitely better. I agree with your reasoning.

**Comment by kukreja-vikramaditya** [comment]
@romani 
So I would only need to update the property name to `alwaysDemandTrailingComma` and update the documentation of the ArrayTrailingComma check?
>The check demands a comma at the end if neither left nor right curly braces are on the same line as the last element of the array

**Comment by kukreja-vikramaditya** [comment]
@jsotuyod Do you think we should have special processing for anonymous arrays (like only check them or skip them) ? Or i guess it would be clear more once reports are generated .

**Comment by jsotuyod** [comment]
@kukreja-vikramaditya I think all arrays should be treated equally.

**Comment by romani** [comment]
yes, lets treat all arrays in the same way.

@kukreja-vikramaditya , name of property is good.
if that property is TRUE all you need to make sure that there is `,` at the end of array AST.
Please close your current PR, as it should seriously redone.

**Comment by romani** [comment]
@kukreja-vikramaditya , please review why we do not report violation on `return new int[] { 0 };` , html documentation clain there should be violation. Smth is not right code or documentation.

**Comment by kukreja-vikramaditya** [comment]
On it

**Comment by Gaurav-Punjabi** [comment]
@romani if @kukreja-vikramaditya is not working on this issue, can I take it? 
Since it hasn't been touched for a long time?

**Comment by strkkk** [comment]
@Gaurav-Punjabi yes, you can take this issue

**Comment by Gaurav-Punjabi** [comment]
I'm on it.

**Comment by rnveach** [comment]
Fix was merged
